testing/python/metafunc.py

```diff
@@ -1486,6 +1568,33 @@ def test_foo(x):
             ]
         )
 
+    def test_parametrize_module_level_test_with_class_scope(
+        self, pytester: Pytester
+    ) -> None:
+        pytester.makepyfile(
+            """
+            import pytest
+
+            @pytest.fixture
+            def item(request):
+                return request._pyfuncitem
+
+            fixturedef = None
+
+            @pytest.mark.parametrize("x", [0, 1], scope="class")
+            def test_1(item, x):
+                global fixturedef
+                fixturedef = item._fixtureinfo.name2fixturedefs['x'][-1]
```

Comment:
Can you think of a way to test the node fixturedef caching business (in particular the scope=class without class case) in a purely functional test, i.e. without peeking into the pytest internals? I did try myself, but failed.

If there is no user-visible difference, it is better not to test at all, because if the test breaks it doesn't necessarily mean that it's bad.

It will still be good to cover the `if node is None` branch, it's currently not executed at all by any test.

Reply:
I couldn't either. But isn't the case that the tests like this which check pytest internals, help us to assert the pytest works as "how" we expect? Don't they help us to make sure it's the solution in our mind that yields the correct output not a random thing or another part of the code? Such tests might also help developers to remember how they've implemented things.

Regarding `if node is None`, it takes place when there's function-scoped direct parametrization in which we do not save any fixturedef to nodes since each function has its dedicated fixturedef. A lot of tests use this kind of parametrization e.g. `TestMetafunc::test_parametrize_onearg`. You mean a dedicated one? E.g. `test_function_scoped_direct_parametrize`?

Reply:
You're right. I try to update the test. And what about `if node is None`?

Reply:
> And what about if node is None?

In #11256 I added the most basic test for this (just that it doesn't crash), it's sufficient for now.

It would be nice to have a unit-test that the fixturedef is indeed reused but I wouldn't spend too much time on it.

Reply:
I changed the test. Is it appropriate now?